[WIP] Multi NFT feature

kalwalt opened the pull request:
# Multi NFT feature
Starting to add the multi NFT feature as in https://github.com/webarkit/jsartoolkit5/pull/7
This PR will add the multi NFT feature to JsartoolkitNFT in Typescript.

## List of Features
- multi NFT markers with a new **loadNFTMarkers** function
- updated **getNFData** function: it return an object with width, height and dpi for every NFT markers.
- github action to commit the generated build and dist files.
- updated WebARKitLib and separate detection from tracking in emscripten/ARToolKitJS.cpp. (using WebARKitLib from dev branch)

### Things to fix:
- [x] other examples (basic, multi, and simple ES6 example): they need changes where getNFTData is used.
- [x] add the new getNFTData in `artoolkitNFT.api.js`
- [x] try to fix error in comment https://github.com/webarkit/jsartoolkitNFT/pull/102#issuecomment-1001250124.
- [x] clean the code
- [x] github action script: commit the created build libs
- [x] final check

kalwalt committed:
new getNFTData method

kalwalt committed:
fix for wrong id increment

kalwalt committed:
webarkitlib multi-nft

kalwalt committed:
moving detection to detectNFTMarker
- see PR https://github.com/webarkit/jsartoolkit5/pull/7

kalwalt committed:
linting code

kalwalt committed:
Merge branch 'feature-multi-nft-ts' into testing-nft-export

kalwalt committed:
Merge branch 'testing-nft-export' into feature-multi-nft-ts

kalwalt committed:
rebuilding libs after merging

kalwalt committed:
new addNFTMarkers function

kalwalt committed:
addNFTMarkers to artoolkitNFT.api.js

kalwalt committed:
start to implements addNFTMarkers method
- not load yet the NFT marker

kalwalt committed:
now is loaded, but twice
- for loop to be fixed

kalwalt commented on src/ARToolkitNFT.ts:
```diff
@@ -269,32 +269,35 @@ export default class ARToolkitNFT {
-    }*/
+    let prefixes: any = [];
+    let vec;
     for (var i = 0; i < urls.length; i++) {
```
i think the loop should be a bit different, it load twice the marker because the url lenght is 1 (in this case of course that we load ONLY one marker) so it load the 0 and the 1 url, but it should load only the 0 address of the array.

kalwalt replied:
i don't think that this is the root of the issue.

kalwalt committed:
testing output

kalwalt committed:
maybe this is more correct

kalwalt commented on src/ARToolkitNFT.ts:
```diff
@@ -271,6 +271,7 @@ export default class ARToolkitNFT {
     let out;
     let prefixes: any = [];
     let vec;
+    prefixes.push(targetPrefix);
```
now it is loaded only one time, but i need to test.

kalwalt replied:
it seems that the marker is detected but not tracked.

kalwalt replied:
with the latest changes https://github.com/webarkit/jsartoolkitNFT/pull/102/commits/c3a21d972533b8e1c7c490fe0aa6b33b40ee37e5, the marker is also tracked!

kalwalt committed:
return markerIds

kalwalt committed:
fix for id

kalwalt committed:
marker now is detected and tracked

kalwalt committed:
new loadNFTMarker and loadNFTMarkers
-  removed old methods

kalwalt commented:
Last steps: i need to add a multi NFT markers example and it will be ready to be merged! :smile:

kalwalt committed:
multi examples

kalwalt commented:
Multi example with artoolkit.api.js works, it can be loaded and track the 3 NFT markers, but the mutli ES6 example fail to load all the NFT markers. Probably a trivial mistake, i hope to fix it soon.

kalwalt committed:
partially solved

kalwalt committed:
new private _storeNFTMarkers function W.I.P.

kalwalt committed:
missed delegate method

kalwalt committed:
testing an old way

kalwalt committed:
small changes

kalwalt committed:
other tests

albjeremias commented:
how is it going? :)

kalwalt commented:
> how is it going? :)

Hi @albjeremias i haven't continued on this, unfortunately i can't  yet develop typescript version. i haven't yet solved. But at least it works for the javascript version. I have also implemented the[ ARnft multi NFT ](https://github.com/webarkit/ARnft/pull/199)version if you can interest you. It's another approach, becuase it create a different WebWorker for every NFT marker. Instead with this PR #102 you can load multi NFT in the same WebWorker.
If you are interested i could merge the code that works.

albjeremias commented:
so.. i came back to the starting point :))) now i read you better... yes i dont think that it is a very good approach, at least for me it did not work. I'd like to known in which state this is.. and if I can try to help somehow. anyway i will test this branch

kalwalt commented:
> so.. i came back to the starting point :))) now i read you better... yes i dont think that it is a very good approach, at least for me it did not work. I'd like to known in which state this is.. and if I can try to help somehow. anyway i will test this branch

I haven't worked again on. You will see that i was trying different things with the ES6 code. The code that is working is the standard API that load the artoolkitNFT.min.js lib init the worker. You can test the [multi example](https://github.com/webarkit/jsartoolkitNFT/pull/102/files#diff-b493c124a4de51c5ecda09a5fff8b39e5d849711ddade9063830cac94e85110a). 
The main problem for the ES6 version of the lib was converting the javascript code to Typescript, maybe you can give me some help in this sense.

kalwalt committed:
fix for Utils
- modified loadnftMarkers

kalwalt committed:
NFT marker now is detected and tracked!

kalwalt commented:
@albjeremias with the latest commit https://github.com/webarkit/jsartoolkitNFT/pull/102/commits/f19446c28994213a6190e4f0b30a110e3816ca61 the pinball image can be detected in the [ARtoolkitNFT_multi_ES6_example.html](https://github.com/webarkit/jsartoolkitNFT/blob/feature-multi-nft-ts/examples/ARtoolkitNFT_multi_ES6_example.html). Still i need to make some modifications to the code, but i'm quite sure that will work for multi NFT markers.
I can't say when it will be ready, because i'm very busy during these days, but maybe before the end of the year?!? :thinking:

kalwalt committed:
adding ome models for the other NFT markers

kalwalt committed:
simplyfing the code

kalwalt committed:
for loop for tracking ids

kalwalt committed:
improved workers

kalwalt commented:
**getNFTData** only get the nft data from the first id, it should instead retrieve the data of all markers. I will work on this.

kalwalt committed:
new getNFTData function

kalwalt commented:
Working on a new getNFTData function not yet solved. I will see when i can continue on this...

kalwalt committed:
improving getNFTData

kalwalt committed:
Merge branch 'dev' into feature-multi-nft-ts
- some libs need rebuilding

kalwalt committed:
rebuilding libs for safety

kalwalt committed:
setting nodejs version

kalwalt committed:
getNFTData at index

kalwalt commented:
I think that this PR is on the good road to be merged. I have to solve some key points but these will not require so much effort. :smile:

kalwalt committed:
auto positioning for the sphere (pinball marker)

kalwalt committed:
cube and cone repositioning

kalwalt commented:
Fixed the `ARToolkitNFT_ES6_example`. I noticed that i get this error:  
```
artoolkitNFT_ES6_wasm.js:9 [info] Page[0]  pre: 51, aft: 51, error = 0.630476
put_char @ artoolkitNFT_ES6_wasm.js:9
write @ artoolkitNFT_ES6_wasm.js:9
write @ artoolkitNFT_ES6_wasm.js:9
doWritev @ artoolkitNFT_ES6_wasm.js:9
_fd_write @ artoolkitNFT_ES6_wasm.js:9
$func990 @ 001da356:0x395df
$func552 @ 001da356:0x1d7c2
$func211 @ 001da356:0x6a24
$func56 @ 001da356:0x182f
$func1197 @ 001da356:0x5dba8
$func1158 @ 001da356:0x4fb1f
detectNFTMarker @ VM25:8
detectNFTMarker @ ARControllerNFT.ts:323
process @ ARControllerNFT.ts:258
process @ artoolkitNFT_multi_ES6.worker.js:66
self.onmessage @ artoolkitNFT_multi_ES6.worker.js:12
artoolkitNFT_ES6_wasm.js:9 [info] Page[1]  pre: 51, aft: 51, error = 0.630476
put_char @ artoolkitNFT_ES6_wasm.js:9
write @ artoolkitNFT_ES6_wasm.js:9
write @ artoolkitNFT_ES6_wasm.js:9
doWritev @ artoolkitNFT_ES6_wasm.js:9
_fd_write @ artoolkitNFT_ES6_wasm.js:9
$func990 @ 001da356:0x395df
$func552 @ 001da356:0x1d7c2
$func211 @ 001da356:0x6a24
$func56 @ 001da356:0x182f
$func1197 @ 001da356:0x5dba8
$func1158 @ 001da356:0x4fb1f
detectNFTMarker @ VM25:8
detectNFTMarker @ ARControllerNFT.ts:323
process @ ARControllerNFT.ts:258
process @ artoolkitNFT_multi_ES6.worker.js:66
self.onmessage @ artoolkitNFT_multi_ES6.worker.js:12
artoolkitNFT_ES6_wasm.js:9 [info] Page[2]  pre: 51, aft: 51, error = 0.630476
```

kalwalt committed:
fixing getNFTData in the standard api

kalwalt commented:
> Fixed the `ARToolkitNFT_ES6_example`. I noticed that i get this error:
> 
> ```
> artoolkitNFT_ES6_wasm.js:9 [info] Page[0]  pre: 51, aft: 51, error = 0.630476
> put_char @ artoolkitNFT_ES6_wasm.js:9
> write @ artoolkitNFT_ES6_wasm.js:9
> write @ artoolkitNFT_ES6_wasm.js:9
> doWritev @ artoolkitNFT_ES6_wasm.js:9
> _fd_write @ artoolkitNFT_ES6_wasm.js:9
> $func990 @ 001da356:0x395df
> $func552 @ 001da356:0x1d7c2
> $func211 @ 001da356:0x6a24
> $func56 @ 001da356:0x182f
> $func1197 @ 001da356:0x5dba8
> $func1158 @ 001da356:0x4fb1f
> detectNFTMarker @ VM25:8
> detectNFTMarker @ ARControllerNFT.ts:323
> process @ ARControllerNFT.ts:258
> process @ artoolkitNFT_multi_ES6.worker.js:66
> self.onmessage @ artoolkitNFT_multi_ES6.worker.js:12
> artoolkitNFT_ES6_wasm.js:9 [info] Page[1]  pre: 51, aft: 51, error = 0.630476
> put_char @ artoolkitNFT_ES6_wasm.js:9
> write @ artoolkitNFT_ES6_wasm.js:9
> write @ artoolkitNFT_ES6_wasm.js:9
> doWritev @ artoolkitNFT_ES6_wasm.js:9
> _fd_write @ artoolkitNFT_ES6_wasm.js:9
> $func990 @ 001da356:0x395df
> $func552 @ 001da356:0x1d7c2
> $func211 @ 001da356:0x6a24
> $func56 @ 001da356:0x182f
> $func1197 @ 001da356:0x5dba8
> $func1158 @ 001da356:0x4fb1f
> detectNFTMarker @ VM25:8
> detectNFTMarker @ ARControllerNFT.ts:323
> process @ ARControllerNFT.ts:258
> process @ artoolkitNFT_multi_ES6.worker.js:66
> self.onmessage @ artoolkitNFT_multi_ES6.worker.js:12
> artoolkitNFT_ES6_wasm.js:9 [info] Page[2]  pre: 51, aft: 51, error = 0.630476
> ```

It's not an error it's normal behaviour; but maybe should print only one time the line
`[info] Page[0]  pre: 51, aft: 51, error = 0.630476`
comparing example from https://www.webarkit.org/examples/multi_wasm_nft.html

kalwalt committed:
solved issue with multi Page printing
- see https://github.com/webarkit/jsartoolkitNFT/pull/102#issuecomment-1001250124
- WebARKitLib realigned to https://github.com/webarkit/WebARKitLib/compare/feature-multi-nft

kalwalt commented:
Last commit solved the issue (see comments before): the WebarKitLib wasn't pointing to https://github.com/webarkit/WebARKitLib/compare/feature-multi-nft. For same reason i was using instead the master branch.
I need to do the last check and that PR will be ready to be merged. :slightly_smiling_face:

kalwalt committed:
small fixes

kalwalt committed:
cleaning the code

kalwalt committed:
production build

kalwalt commented:
I have to do last changes, i think it will be merged at the beginning of the new year, can't do now... :smile:

kalwalt committed:
commit the new files in the github action

github-actions[bot] committed:
new build files from action

kalwalt commented:
The action github script commit the created files, so this is the final step of this PR. I'm updating the topic thread summarizing the features added.

kalwalt committed:
cleaning the code  and rebuilding

kalwalt committed:
Merge branch 'feature-multi-nft-ts' of https://github.com/webarkit/jsartoolkitNFT into feature-multi-nft-ts

github-actions[bot] committed:
new build files from action

kalwalt committed:
using WebARKitLib dev branch

kalwalt committed:
Merge branch 'feature-multi-nft-ts' of https://github.com/webarkit/jsartoolkitNFT into feature-multi-nft-ts

github-actions[bot] committed:
new build files from action
